src/Compilers/Core/Portable/Operations/Operation.Enumerable.cs

```diff
@@ -0,0 +1,116 @@
+            private readonly Operation _operation;
+
+            internal Enumerable(Operation operation)
+            {
+                _operation = operation;
+            }
+
+            public Enumerator GetEnumerator() => new Enumerator(_operation);
+
+            public ImmutableArray<IOperation> ToImmutableArray()
+            {
+                switch (_operation)
+                {
+                    case NoneOperation { Children: var children }:
+                        return children;
+                    case InvalidOperation { Children: var children }:
+                        return children;
+                    case var _ when !GetEnumerator().MoveNext():
+                        return ImmutableArray<IOperation>.Empty;
+                    default:
+                        var builder = ArrayBuilder<IOperation>.GetInstance();
+                        foreach (var child in this)
+                        {
+                            builder.Add(child);
+                        }
+                        return builder.ToImmutableAndFree();
+                }
+            }
+
+            IEnumerator<IOperation> IEnumerable<IOperation>.GetEnumerator() => this.GetEnumerator();
+            IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
+        }
+
+        /// <summary>
+        /// Implements a struct-based enumerator for <see cref="Operation"/> nodes, using a slot-based system that tracks
+        /// the current slot, and the current index in the slot if the current slot is an immutable array. This type is not hardened
+        /// to default instances, and will null reference in these cases. Implementation of the <see cref="Enumerator.MoveNext"/> and
+        /// <see cref="Enumerator.Current"/> members are delegated to the virtual <see cref="Operation.MoveNext(int, int)"/> and
+        /// <see cref="Operation.GetCurrent(int, int)"/> methods, respectively.
+        /// </summary>
+        [NonDefaultable]
+        internal struct Enumerator : IEnumerator<IOperation>
+        {
+            private readonly Operation _operation;
+            private int _currentSlot;
+            private int _currentIndex;
+
+            public Enumerator(Operation operation)
+            {
+                _operation = operation;
+                _currentSlot = -1;
+                _currentIndex = -1;
+            }
+
+            public IOperation Current
+            {
+                get
+                {
+                    Debug.Assert(_operation != null && _currentSlot >= 0 && _currentIndex >= 0);
+                    return _operation.GetCurrent(_currentSlot, _currentIndex);
+                }
+            }
+
+            public bool MoveNext()
+            {
+                switch (_operation.MoveNext(_currentSlot, _currentIndex))
+                {
+                    case (false, _, _):
+                        _currentSlot = int.MinValue;
+                        _currentIndex = int.MinValue;
```

Comment:
This doesn't look right, given the discussion around MoveNext implementation

Reply:
Yep, I've already corrected this locally.

---
In reply to: [541181111](https://github.com/dotnet/roslyn/pull/49804#discussion_r541181111) [](ancestors = 541181111)